### DISCUSSION TITLE
Version targeting for 6.x

### DISCUSSION THREAD

**Discussion by starkos** [premake/premake-core#1698] [Show and tell]
**TL;DR:** 6.x has a new approach for targeting exported project file features to specific toolset versions. ICYMI.

We've lately seen both Microsoft and Apple introducing potentially breaking project file changes in minor version updates of their tools. The way we set up v5.x with "one exporter per major version" there isn't a whole lot we can do about that: we target the current latest version for 2019, and folks who are pinned to an older version are…just out of luck, really.

The work on 6.x has reached the point where I'm exporting some version specific stuff; I figured I'd take a stab and this and set things up right from the start. I've introduced a [new Version module](https://github.com/starkos/premake-core/tree/6.x/core/modules/version), with some [basic documentation here](https://github.com/starkos/premake-core/tree/6.x/website/docs/version).

When using Premake6 to target Visual Studio, you can now fully control the version you're targeting.

```sh
# Target the latest that Premake supports; currently our most complete VS 2019
premake6 vstudio

# Target a specific "named" version; like `premake5 vs2017`
premake6 vstudio=2017

# Now you can also specify an exact build
premake6 vstudio=16.4.31429.391
```

The Visual Studio exporter code can now take advantage of this to target specific versions ranges:

```lua
if vstudio.targetVersion:is('2017') then ... end

if vstudio.targetVersion:isAtLeast('16.4') then ... end

if vstudio.targetVersion:isNoLaterThen('16.4') then ... end
```

A few of those are still "coming soon"…I'll fill them in as soon as I hit a place that needs them. Here's [a value-mapping example from the current code](https://github.com/starkos/premake-core/blob/231c16e73ed65a2a89b1823abd4fb6c5fcfef662/modules/vstudio/src/vcxproj.lua#L297-L303).

Feedback welcome, especially if I've missed something in the approach!
